Issue 46 temp

silver-1347 opened the pull request:
要求を満たすようなテストを自分なりに書いて、そのテストを通るような実装をしたのですが、実際にブラウザでチェックすると、テストでは出ないエラーなどが発生しています。

テストに不足があると思うのでそこのレビューだけでもお願いしたいです。

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

participant_settings_testへメールアドレス更新のテストを追加

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

送信メールのテストを追加
new_email -> emailの順を踏んでメールを更新するようにテストを修正

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

email_update_tokenとnew_emailを追加

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

テストを微修正

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

participantにchanging_emailを追加

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

テストにchanging_emailに関するテストを追加

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

new_emailのバリデーションの追加とemailのユニーク性についてnew_emailとも被らないようにバリデーションを修正
participants_controllerのupdateにnew_emailの更新を実装(メール送信はまだ)
viewにメール更新フォームを追加

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

テストの修正

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

participantにメール更新トークンに対応するダイジェストを入れるカラムを追加
accountにメール更新トークンとダイジェストを作成するメソッドと認証するメソッドを追加
participants_controllerのupdateでのトークンとダイジェスト作成をするように修正

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

メールアドレス更新のメール送信関係のファイルを追加
routesにemail_updateを追加

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

routeから呼び出されるemail_updateに対応するアクションを持つparticipants_email_update_controllerを追加

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

メール更新の本体をコントローラに実装
emailとnew_emailのバリデーションをそれぞれに被りがないようにするものに修正
new_emailでparticipantを探せるメソッドを追加
メール内容のミスを修正
メール更新部分のテストを微修正

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)
メールアドレスに有効期限を設けた
　メールアドレスの有効期限のテストを実装
　メールアドレス更新トークンを作成したタイミングを記録するカラムを追加
　有効期限切れかを確かめるメソッドを追加
　メール文面に有効期限について記載

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

view部分でのミスを修正

okabi commented on app/views/participants/_form.html.erb:
```diff
@@ -12,7 +12,7 @@
   <% end %>
-  <% if type == 'all' %>
+  <% if type == 'all' || type == 'email_update'%>
     <%= form_email_field f, :email, 'メールアドレス', description: 'ログインや連絡に用いるため、普段お使いのメールアドレスをご入力ください。' %>
```
ここの `name` が `:email` になっているため、email_update 時にコントローラがパラメータを受け取れていないようです。
all (アカウント登録)のときは `:email` 、email_update のときは `:new__email` にすると良さそうです。
ただ、その場合現在設定されているメールアドレスがフォームの既存値にならずユーザーに不親切な感じなので、
ケアしてもらえるとありがたいです。

okabi reviewed (changes requested):
テストは良いと思います 😄 
ざっと動作確認しましたが、レビューした点を直すと想定通りの動作をしてそうです。
コードについてはまだ読み込んでないので、またレビュー依頼もらったときにやります。

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

viewでemail_updateの部分でnew_emailに入力されるように修正した。既定値として現アドレスを表示されるようにすることはまだできていない。

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

descriptionで現在設定されているメールアドレスを表示するようにした

silver-1347 commented:
ブラウザでも動くようにできました。レビューお願いします。

バリデーション周りが自信がないです。
+ `email` と`new_email` で相互に被りがないかチェックするようにしたのですがだいぶ怪しい実装になっています。 `particilant.validate_email_uniqueness` と `particilant.validate_new_email_uniqueness` です。
+ メール更新のフラグとして `changing_email` を用意して `new_email` のバリデーションをするかどうかを決めていますが、これによってバリデーションに漏れが出たりしていないか少し自信がないです。(一応テスト通ってるので大丈夫かとは思っていますが…)

結構いろんなファイルを変更したので大変かと思いますが、よろしくお願いします。

yegawa commented on app/controllers/participants_controller.rb:
```diff
@@ -44,13 +44,27 @@ def update
+    elsif @type == 'email_update'
+      @participant.changing_email = true
+      @participant.create_email_update_token_and_digest
+      participant_params = params.require(:participant).permit(:new_email)
```
好みの問題なんだろうけど、この辺の `params.require(:participant).permit(:hoge)` たちが
1行すごい長くなってくるのが伺えるし情報多いから 
`password_params` とか `email_update_params` `profile_params` とかそんな感じで
別メソッドにまとめてやるのもアリだと思う。

yegawa commented on app/controllers/participants_controller.rb:
```diff
@@ -44,13 +44,27 @@ def update
+        flash.now[:danger] = 'メールアドレスの更新に失敗しました。'
+        @participant.changing_email = false
+        @participant.email_update_token = nil
+        @participant.email_update_digest = nil
```
これで更新できたっけ？
`@participant.update_attributes = { :changing_email = false, ...}`
って感じにしないと行けない気がする。
ここに来るときはおそらくバリデーションで弾かれたときだから、このupdate_arrtibutesは成功するはず
（さらにエラーチェックして500に飛ばすとかいうのを書く必要はさすがに無いと思いたい）

yegawa commented on app/controllers/participants_email_update_controller.rb:
```diff
@@ -0,0 +1,23 @@
+    token = params[:t]
+    participant = Participant.find_by_new_email(new_email)
+    if participant.email_update_authenticated?(token) && !participant.email_update_token_expired?
+      participant.update(email: new_email,new_email: nil,changing_email: false)
```
他との統一感がないのと読みにくいので 
`participant.update(email: new_email, new_email: nil, changing_email: false)`
としたいけど、一行が長くなるのは避けたい、
かと言って改行入れるのも変。
こういうときどうすりゃいいんやろうね。

yegawa commented on app/controllers/participants_email_update_controller.rb:
```diff
@@ -0,0 +1,23 @@
+    def failed_email_update
+      flash[:danger] = "メールアドレスの更新に失敗しました"
+      redirect_to participant_login_url
+    end
```
このメソッド化 好き

yegawa commented on app/controllers/participants_email_update_controller.rb:
```diff
@@ -0,0 +1,23 @@
+    new_email = params[:e]
+    token = params[:t]
+    participant = Participant.find_by_new_email(new_email)
+    if participant.email_update_authenticated?(token) && !participant.email_update_token_expired?
```
この2つの条件ひっくるめたメソッドを
`app/models/concerns/account.rb` あたりに作ったほうがmemberに実装する時ミスらなさそう。

yegawa commented on app/mailers/participant_mailer.rb:
```diff
@@ -20,4 +20,9 @@ def password_reset(participant)
+  def email_update(participant)
+    @participant = participant
+    mail to: @participant.new_email, subject: "メールアドレスの更新"
+  end
```
（この辺もそのうち `concerns` にまとめる作業しないとなあ…）

yegawa commented on app/models/participant.rb:
```diff
@@ -52,10 +54,18 @@ def validate_birth
   # アクティブ被験者のメールアドレスのユニーク性バリデーション
   def validate_email_uniqueness
-    participant = Participant.find_by(deactivated: false, email: email.downcase)
+    participant = Participant.find_by("deactivated == ? and (email == ? or (changing_email == ? and new_email == ?))",false,email.downcase,true,email.downcase)
```
Rails 5からSQLべた書きせずにorも書けるようになったらしい！
http://qiita.com/QUANON/items/0033a5adb76d0cb963cd

俺も書いたこと無いから書き方分からんけど、
適当に書きながらrails sのログで実行されてるSQL文と比較して書くと良いと思うよ！

yegawa commented on app/models/participant.rb:
```diff
@@ -52,10 +54,18 @@ def validate_birth
+  # メール更新時のメールアドレスのユニーク性を検証する
+  def validate_new_email_uniqueness
+    if !new_email.nil?
+      participant = Participant.find_by("deactivated == ? and (email == ? or (changing_email == ? and new_email ==   ?))",false,new_email.downcase,true,new_email.downcase)
```
orに関して、同上。

yegawa commented on app/models/participant.rb:
```diff
@@ -52,10 +54,18 @@ def validate_birth
 
+  # メール更新時のメールアドレスのユニーク性を検証する
+  def validate_new_email_uniqueness
+    if !new_email.nil?
```
これ `if new_email` でイケルと思う。

okabi replied:
`return if new_email.nil?` としてインデントが深くならないようにしたいかな。

yegawa commented on app/models/participant.rb:
```diff
@@ -82,6 +92,10 @@ def self.find_by_email(email)
 
+  def self.find_by_new_email(new_email)
+    Participant.find_by(new_email: new_email, deactivated: false)
+  end
```
良い拡張。
将来concernsに行くかも。

okabi replied:
メソッド名がちょっと危ない気がします( `new_email` だけで絞ってるように見える)。
アクティブ被験者のみを取っていることをメソッド名から推定できるようにしたい。

yegawa commented on test/integration/participants_settings_test.rb:
```diff
@@ -100,4 +120,34 @@ def update_password(password, password_confirmation)
+    participant.reload
+    assert_equal email, participant.email
+    # トークンの期限が切れている場合
+    participant.update(set_email_update_at: participant.set_email_update_at - 86401)
```
`24.hours.ago(participant.set_email_update_at).ago(1)` 
とか
`24.hours.ago(1.secondes.ago(participant.set_email_update_at)` 
にするか迷いどころ（ago二回はキモいもんなあ…）

yegawa commented on test/integration/participants_settings_test.rb:
```diff
@@ -100,4 +120,34 @@ def update_password(password, password_confirmation)
+    get email_update_url(t: participant.email_update_token, e: new_email)
+    participant.reload
+    assert_equal email, participant.email
+    participant.update(set_email_update_at: participant.set_email_update_at + 86401)
```
ago使うならsince

yegawa reviewed (changes requested):
ボヤキとかも多いですが一応レビューしました。
コード上でしか見ていないのであしからず…。

必須
・ `@participant.update_attributes = { :changing_email = false, ...}` にする必要があるかチェック
・ `participant.email_update_authenticated?(token) && !participant.email_update_token_expired?` をまとめる
・ モデルの検索でor使ってるのを書き換える

で後は適当に判断してください

okabi commented on app/controllers/participants_email_update_controller.rb:
```diff
@@ -0,0 +1,23 @@
+  include AccountsCommon
+
+  def update
+    new_email = params[:e]
```
`new_email = params[:e].downcase` にしときたい

okabi commented on test/mailers/participant_mailer_test.rb:
```diff
@@ -66,4 +66,20 @@ class ParticipantMailerTest < ActionMailer::TestCase
+  test "email_update" do
+    participant = participants(:one)
+    participant.new_email = "update@example.com"
+    # participant.changing_email = true
```
特に意図がないならコメントアウトしたものは消しといて欲しいです。

okabi commented on app/models/participant.rb:
```diff
@@ -52,10 +54,18 @@ def validate_birth
 
   # アクティブ被験者のメールアドレスのユニーク性バリデーション
   def validate_email_uniqueness
```
なるほど、メールアドレス更新時に新規登録が来た場合対策か！
よく気付いてくれたと思います。素晴らしい。

okabi commented on db/migrate/20170408160354_add_email_updates_to_participants.rb:
```diff
@@ -0,0 +1,9 @@
+class AddEmailUpdatesToParticipants < ActiveRecord::Migration[5.0]
+  def change
+    add_column :participants, :email_update_token, :string
```
トークンをDBに保存するのは危ない！！
保存するのは digest のみ。
ちょっと抽象度高くなってて見づらいけど、participant_password_resets_controller.rb とか見て欲しい。

okabi commented on db/migrate/20170408160354_add_email_updates_to_participants.rb:
```diff
@@ -0,0 +1,9 @@
+    add_column :participants, :email_update_digest, :string
+    add_column :participants, :set_email_update_at, :datetime
+    add_column :participants, :new_email, :string
+    add_column :participants, :changing_email, :boolean, default: false
```
`changing_email` ってDBに保存しておく必要あるのかな？
`set_email_update_at` だけで判定できる気がします。

okabi reviewed (changes requested):
ざっくりとレビューしました。
セキュリティ的にまずいところがあったので、そこは必須修正でお願いしたいです。

silver-1347 committed:
[#ref 46] メールアドレスを更新できるようにする(被験者)

DB上からemail_update_tokenのカラムを削除
メールアドレス更新失敗した場合の処理の部分を修正
コメントアウトしたコードの削除
changing_emailになったので不要なので削除

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

participantのemailとnew_emailのバリデーションでのfind_byの書き方を修正
participant.email_update_authenticated?で期限切れもチェックするようにして条件式の部分をまとめた

その他の指摘点を修正
+ email_update_tokenを作るタイミングを変更
+ find_by_new_emailをfind_active_by_new_emailに名前変更
+ new_emailをdowncaseして入れるように修正
+ テストでago,sinceを使うように修正
+ new_emailのvalidationでインデントが深くならないように修正

silver-1347 commented:
レビューいただいた部分について必須事項の全てと、指摘点のいくつかを修正しました。

yegawa commented on app/models/participant.rb:
```diff
@@ -55,32 +55,31 @@ def validate_birth
-      if participant && (id.nil? || participant.id != id)
-        errors.add(:email, "は既に登録済みです")
-      end
+    return new_email.nil?
```
この書き方だと下の方到達しない気がするんだけどどうでしょうか？
return hoge if fuga
なら行けると思うけど
return false if new_email.nil?
とかかね？

silver-1347 replied:
あ…なんか色々頭回ってなかったみたいです…

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

new_emailのバリデーションの修正

okabi commented on app/models/participant.rb:
```diff
@@ -52,8 +54,33 @@ def validate_birth
   # アクティブ被験者のメールアドレスのユニーク性バリデーション
   def validate_email_uniqueness
+    # 現在登録されているメールアドレスとのユニーク性
     participant = Participant.find_by(deactivated: false, email: email.downcase)
```
`Participant.find_active_by_email` を定義して使いたい。
`validate_new_email_uniqueness` の中身も同様。
(できれば他に同様の処理をしている部分を、新しく定義した関数で置き換えてもらえると助かります)

okabi commented on app/models/participant.rb:
```diff
@@ -52,8 +54,33 @@ def validate_birth
+    end
+
+    # 更新予定のメールアドレスとのユニーク性
+    participant = Participant.find_by(deactivated: false, new_email: email.downcase)
```
せっかく定義したし、 `Participant.find_active_by_new_email` を使いたい。
`validate_new_email_uniqueness` の中身も同様。
(できれば他に同様の処理をしている部分を、新しく定義した関数で置き換えてもらえると助かります)

okabi reviewed (commented):
あとで動作確認しようかと思うけど、とりあえずコードだけ見てぼやきみたいなレビュー。

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

find_active_by_emailを新しく追加
find_active_by_email/new_emailをバリデーションで利用するように修正

okabi commented on app/models/participant.rb:
```diff
@@ -52,8 +54,33 @@ def validate_birth
+
+    # 更新予定のメールアドレスとのユニーク性
+    participant = Participant.find_active_by_new_email(email.downcase)
+    if participant && (id.nil? || participant.id != id) && participant.email_update_token_expired?
```
ごめん、さっき見逃してたんだけど `participant.email_update_token_expired?` って条件として合ってますか？
更新予定のアドレスと被ってほしくないので、 `!participant.email_update_token_expired?` のような気がします。
( `validate_new_email_uniqueness` の中身も同様)

okabi commented on app/models/participant.rb:
```diff
@@ -52,8 +54,33 @@ def validate_birth
+    return true if new_email.nil?
+    # 現在登録されているメールアドレスとのユニーク性
+    participant = Participant.find_active_by_email(new_email.downcase)
+    if participant && (id.nil? || participant.id != id)
```
動作確認しました。
動作的には一点だけ修正してほしい点があって、現在設定しているメールアドレスと同じアドレスで更新しようとした時に
バリデーションで弾いてほしいです(それに関するテストもほしいかな)。
おそらく `if participant` で良いかな？

okabi commented on app/models/participant.rb:
```diff
@@ -52,8 +54,33 @@ def validate_birth
+    # 現在登録されているメールアドレスとのユニーク性
+    participant = Participant.find_active_by_email(new_email.downcase)
+    if participant && (id.nil? || participant.id != id)
+      errors.add(:email, "は既に登録済みです")
```
ごめん、見逃してたけどこの辺 `:email` から `:new_email` にしないと行けないような気がする。

okabi reviewed (commented):
動作確認しました。さっき見逃してた点もレビューしました。。

silver-1347 committed:
[ref #46] メールアドレスを更新できるようにする(被験者)

バリデーションで条件がおかしかったところを修正
現在登録しているアドレスでメールアドレスを更新しようとするのを防ぐように修正
すでに登録済みの場合に表示するエラーメッセージのところをnew_emailで表示するように修正

okabi reviewed (approved):
良さそう 👍

okabi committed:
Merge branch 'develop' into issue-46-temp
